sway-fmt-v2/src/config/manifest.rs

```diff
@@ -0,0 +1,170 @@
+        Self {
+            whitespace: opts
+                .whitespace
+                .as_ref()
+                .map(Whitespace::from_opts)
+                .unwrap_or_default(),
+            imports: opts
+                .imports
+                .as_ref()
+                .map(Imports::from_opts)
+                .unwrap_or_default(),
+            ordering: opts
+                .ordering
+                .as_ref()
+                .map(Ordering::from_opts)
+                .unwrap_or_default(),
+            items: opts
+                .items
+                .as_ref()
+                .map(Items::from_opts)
+                .unwrap_or_default(),
+            lists: opts
+                .lists
+                .as_ref()
+                .map(Lists::from_opts)
+                .unwrap_or_default(),
+            literals: opts
+                .literals
+                .as_ref()
+                .map(Literals::from_opts)
+                .unwrap_or_default(),
+            expressions: opts
+                .expressions
+                .as_ref()
+                .map(Expressions::from_opts)
+                .unwrap_or_default(),
+            heuristics: opts
+                .heuristics
+                .as_ref()
+                .map(Heuristics::from_opts)
+                .unwrap_or_default(),
+            structures: opts
+                .structures
+                .as_ref()
+                .map(Structures::from_opts)
+                .unwrap_or_default(),
+            comments: opts
+                .comments
+                .as_ref()
+                .map(Comments::from_opts)
+                .unwrap_or_default(),
+        }
+    }
+    /// Given an optional path to a `swayfmt.toml`, read it and construct a `Config`.
+    /// If settings are omitted, those fields will be set to default. If `None` is provided,
+    /// the default config will be applied. If a `swayfmt.toml` exists but is empty, the default
+    /// config will be applied.
+    ///
+    /// At present, this will only return a warning if it catches unusable fields.
+    /// Upon completion, this should give errors/warnings for incorrect input fields as well.
+    ///
+    pub fn from_dir_or_default(config_path: Option<&Path>) -> Result<Self> {
+        let config = ConfigOptions::from_dir_or_default(config_path)?;
+        Ok(config)
+    }
+}
+
+impl ConfigOptions {
+    /// Given an optional path to a `swayfmt.toml`, read it and construct a `Config`.
+    /// If settings are omitted, those fields will be set to default. If `None` is provided,
+    /// the default config will be applied. If a `swayfmt.toml` exists but is empty, the default
+    /// config will be applied.
+    ///
+    /// At present, this will only return a warning if it catches unusable fields.
+    /// Upon completion, this should give errors/warnings for incorrect input fields as well.
+    ///
+    pub fn from_dir_or_default(config_path: Option<&Path>) -> Result<Config> {
```

Comment:
I think we should remove the `Option` around the `&Path`. If a user doesn't have a `&Path`, they should probably just call `Config::default()` directly.

Then, let's rename this to `from_dir` and refactor out a `from_file` function constructor which takes the direct path to the `swayfmt.toml`. Similar to how our `ManifestPath` type has both `from_file` and `from_dir` constructors. We should make sure the doc comments correspond accordingly too.

Also, this function should return `Result<Self>` instead of `Config`. When reading code it can be a bit confusing to see something that looks like a constructor for one type that actually produces a different type. We should do the conversion from `ConfigOptions` to `Config` in the equivalent `Config::from_file` and `Config::from_dir` constructors instead.

Reply:
Okay, I'll try to figure this out today, thanks again so much for all the help and feedback !

Reply:
> I think we should remove the Option around the &Path. If a user doesn't have a &Path, they should probably just call Config::default() directly.

I'm a bit confused on this statement. I figured that the only way we wanted the user to interact with the formatter was via a `swayfmt.toml` and if a path to it doesn't exist then the formatter would just use default. In what way would the user be calling `Config::default()`? 

I was under the assumption that we wanted the config to be finalized before being given to the formatter. This way the only command the user ever needs is `forc fmt`. My apologies for the late reply bit of a busy few days here!

Reply:
Upon further thought, maybe we could have it to where if the toml is absent, we just allow it to fail silently instead of accepting a none value?

Reply:
Okay, I've gone ahead and refactored according to this but my question still stands!

Reply:
> I think we should remove the Option around the &Path. If a user doesn't have a &Path, they should probably just call Config::default() directly.

I think the confusion might come from my use of the word "user" in this sentence. I was referring to the user of the code, (i.e. caller of the function), not the downstream user of rustfmt. Does that make more sense?

---

> and if a path to it doesn't exist then the formatter would just use default (and not return an error, since it is okay for one to be absent).

I think the following is a more elegant solution for `sway-fmt` or `forc-fmt` code that checks for the `Config` file:

```rust
let config = Config::from_dir(path).unwrap_or_default();
```
rather than:
```rust
let config = Config::from_dir_or_default(Some(path));
```
My issue with the latter is that it would require manually checking if the `swayfmt.toml` exists before deciding whether to pass it `Some(path)` or `None`, whereas that check is already done inside `Config::from_dir` anyway.

That said, `unwrap_or_default` probably isn't the right choice here either, as we only want to fallback to the default config if the file doesn't exist, otherwise we want to give some useful error.

Rather than using `anyhow` for `from_dir`'s error type, we might be better off with a custom `ConfigError` using `thiserror`, e.g. something like the following

```rust
/// Errors encountered when loading a `Config` from a `swayfmt.toml` or directory.
#[derive(Debug, Error)]
pub enum ConfigError {
    #[error("failed to parse config: {err}")]
    Deseriallize {
        err: toml::Error,
    }
    #[error("could not find a `swayfmt.toml` in the given directory or its parents")]
    NotFound,
}
```

Then we could match on the result of `from_dir` to handle it properly:

```rust
let config = match Config::from_dir(path) {
    Ok(config) => config,
    Err(ConfigError::NotFound) => Config::default(),
    Err(err) => bail!(err),
};
```

Reply:
This makes much more sense, as you can see I was trying to establish some logic around allowing the absence of the `swayfmt.toml` and ended up with some weird `Option` nonsense. Thanks for clearing this up, will have a commit later today.